### ISSUE TITLE
Maybe interface define in the implementor is not the best practice

### ISSUE THREAD

**Issue by hayden-pan** [MBvisti/weight-tracker-article#3]
https://github.com/golang/go/wiki/CodeReviewComments#interfaces
As the comment above, it's better to return the concrete values, instead of interface. So the better way is define `UserService` struct and return the struct direct, am I right?

**Comment by MBvisti** [comment]
Yes, there is a saying to accept interfaces but return concrete types. I have been working on a follow-up version of the article that shows how to do it that way. However, you will also see this way of doing it some places in the standard library. The way I showed it in the article is a more Java_ish_ way of doing it, since you have to explicit define what implements interfaces in Java, so it's easier to define them like this upfront. That said, the way I prefer to do it now is to return a struct for my services as you say. But both ways is valid in the end :)

**Comment by aaftio** [comment]
Thank you for your answer, I am also come from Java :)

**Comment by alok87** [comment]
Good explanation on the same is given in this article: https://www.appdynamics.com/blog/engineering/three-productive-go-patterns-put-radar/, they call it "Client Side Interface"

TLDR; you want a loose coupling between the consumer(client) and the producer, you use Client Side Interface else you use Producer side interface if you want to keep a tight coupling. 

as @MBvisti said both are valid, and it is based on what you need.
